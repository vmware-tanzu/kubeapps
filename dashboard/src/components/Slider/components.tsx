// Copyright 2019-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

// CODE FROM: https://sghall.github.io/react-compound-slider/#/slider-demos/horizontal
import { GetHandleProps, GetTrackProps, SliderItem } from "react-compound-slider";

// *******************************************************
// HANDLE COMPONENT
// *******************************************************
interface IHandleProps {
  domain: number[];
  handle: SliderItem;
  getHandleProps: GetHandleProps;
}

/* eslint-disable react/prop-types */
export const Handle: React.FC<IHandleProps> = ({
  domain: [min, max],
  handle: { id, value, percent },
  getHandleProps,
}) => (
  <div
    role="slider"
    aria-valuemin={min}
    aria-valuemax={max}
    aria-valuenow={value}
    style={{
      left: `${percent}%`,
      position: "absolute",
      marginLeft: "-11px",
      marginTop: "-6px",
      zIndex: 2,
      width: 24,
      height: 24,
      cursor: "pointer",
      borderRadius: "50%",
      boxShadow: "1px 1px 1px 1px rgba(0, 0, 0, 0.2)",
      backgroundColor: "#34568f",
    }}
    {...getHandleProps(id)}
  />
);

// *******************************************************
// TRACK COMPONENT
// *******************************************************
interface ITrackProps {
  source: SliderItem;
  target: SliderItem;
  getTrackProps: GetTrackProps;
}

export const Track: React.FC<ITrackProps> = ({ source, target, getTrackProps }) => (
  <div
    style={{
      position: "absolute",
      height: 14,
      zIndex: 1,
      backgroundColor: "#7aa0c4",
      borderRadius: 7,
      cursor: "pointer",
      left: `${source.percent}%`,
      width: `${target.percent - source.percent}%`,
    }}
    {...getTrackProps()}
  />
);
