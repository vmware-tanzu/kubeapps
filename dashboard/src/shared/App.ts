import {
  AvailablePackageDetail,
  AvailablePackageReference,
  Context,
  CreateInstalledPackageRequest,
  InstalledPackageReference,
  ReconciliationOptions,
  VersionReference,
} from "gen/kubeappsapis/core/packages/v1alpha1/packages";
import * as url from "shared/url";
import { axiosWithAuth } from "./AxiosInstance";
import { KubeappsGrpcClient } from "./KubeappsGrpcClient";

export const KUBEOPS_ROOT_URL = "api/kubeops/v1";
export class App {
  private static client = () => new KubeappsGrpcClient().getPackagesServiceClientImpl();

  public static async GetInstalledPackageSummaries(
    cluster: string,
    namespace?: string,
    page?: number,
    size?: number,
  ) {
    return await this.client().GetInstalledPackageSummaries({
      context: { cluster: cluster, namespace: namespace },
      paginationOptions: { pageSize: size || 0, pageToken: page?.toString() || "0" },
    });
  }

  public static async GetInstalledPackageDetail(installedPackageRef?: InstalledPackageReference) {
    return await this.client().GetInstalledPackageDetail({
      installedPackageRef: installedPackageRef,
    });
  }

  public static async createInstalledPackage(
    targetContext: Context,
    name: string,
    availablePackageRef: AvailablePackageReference,
    pkgVersionReference: VersionReference,
    values?: string,
    reconciliationOptions?: ReconciliationOptions,
  ) {
    return await this.client().CreateInstalledPackage({
      name,
      values,
      targetContext,
      availablePackageRef,
      pkgVersionReference,
      reconciliationOptions,
    } as CreateInstalledPackageRequest);
  }

  public static async upgrade(
    installedPackageRef: InstalledPackageReference,
    chartNamespace: string,
    availablePackageDetail: AvailablePackageDetail,
    values?: string,
  ) {
    const endpoint = url.kubeops.releases.get(
      installedPackageRef.context?.cluster ?? "",
      installedPackageRef.context?.namespace ?? "",
      installedPackageRef.identifier,
    );
    const { data } = await axiosWithAuth.put(endpoint, {
      appRepositoryResourceName:
        availablePackageDetail.availablePackageRef?.identifier.split("/")[0],
      appRepositoryResourceNamespace: chartNamespace,
      chartName: decodeURIComponent(availablePackageDetail.name),
      releaseName: installedPackageRef.identifier,
      values,
      version: availablePackageDetail.version?.pkgVersion,
    });
    return data;
  }

  public static async rollback(installedPackageRef: InstalledPackageReference, revision: number) {
    const endpoint = url.kubeops.releases.get(
      installedPackageRef.context?.cluster ?? "",
      installedPackageRef.context?.namespace ?? "",
      installedPackageRef.identifier,
    );
    const { data } = await axiosWithAuth.put(
      endpoint,
      {},
      {
        params: {
          action: "rollback",
          revision,
        },
      },
    );
    return data;
  }

  public static async delete(installedPackageRef: InstalledPackageReference, purge: boolean) {
    let endpoint = url.kubeops.releases.get(
      installedPackageRef.context?.cluster ?? "",
      installedPackageRef.context?.namespace ?? "",
      installedPackageRef.identifier,
    );
    if (purge) {
      endpoint += "?purge=true";
    }
    const { data } = await axiosWithAuth.delete(endpoint);
    return data;
  }

  // TODO(agamez): remove it once we return the generated resources as part of the InstalledPackageDetail.
  public static async getRelease(installedPackageRef?: InstalledPackageReference) {
    const { data } = await axiosWithAuth.get<{ data: { manifest: any } }>(
      url.kubeops.releases.get(
        installedPackageRef?.context?.cluster ?? "",
        installedPackageRef?.context?.namespace ?? "",
        installedPackageRef?.identifier ?? "",
      ),
    );
    return data.data;
  }
}
