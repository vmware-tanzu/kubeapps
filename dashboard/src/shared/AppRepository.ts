import { axiosWithAuth } from "./AxiosInstance";
import { IAppRepositoryFilter, ICreateAppRepositoryResponse } from "./types";
import * as url from "./url";

export class AppRepository {
  public static async list(cluster: string, namespace: string) {
    const {
      data: { appRepository },
    } = await axiosWithAuth.get<any>(url.backend.apprepositories.list(cluster, namespace));
    return appRepository;
  }

  public static async get(cluster: string, namespace: string, name: string) {
    const {
      data: { appRepository },
    } = await axiosWithAuth.get<any>(url.backend.apprepositories.get(cluster, namespace, name));
    return appRepository;
  }

  public static async getSecretForRepo(cluster: string, namespace: string, name: string) {
    const {
      data: { secret },
    } = await axiosWithAuth.get<any>(url.backend.apprepositories.get(cluster, namespace, name));
    return secret;
  }

  public static async resync(cluster: string, namespace: string, name: string) {
    const { data } = await axiosWithAuth.post(
      url.backend.apprepositories.refresh(cluster, namespace, name),
      null,
    );
    return data;
  }

  public static async update(
    cluster: string,
    name: string,
    namespace: string,
    repoURL: string,
    type: string,
    description: string,
    authHeader: string,
    authRegCreds: string,
    customCA: string,
    syncJobPodTemplate: any,
    registrySecrets: string[],
    ociRepositories: string[],
    skipTLS: boolean,
    passCredentials: boolean,
    filter?: IAppRepositoryFilter,
  ) {
    const { data } = await axiosWithAuth.put<ICreateAppRepositoryResponse>(
      url.backend.apprepositories.update(cluster, namespace, name),
      {
        appRepository: {
          name,
          repoURL,
          type,
          description,
          authHeader,
          authRegCreds,
          customCA,
          syncJobPodTemplate,
          registrySecrets,
          ociRepositories,
          tlsInsecureSkipVerify: skipTLS,
          passCredentials: passCredentials,
          filterRule: filter,
        },
      },
    );
    return data;
  }

  public static async delete(cluster: string, namespace: string, name: string) {
    const { data } = await axiosWithAuth.delete(
      url.backend.apprepositories.delete(cluster, namespace, name),
    );
    return data;
  }

  // create uses the kubeapps backend API
  // TODO(mnelson) Update other endpoints to similarly use the backend API, removing the need
  // for direct k8s api access (for this resource, at least).
  public static async create(
    cluster: string,
    name: string,
    namespace: string,
    repoURL: string,
    type: string,
    description: string,
    authHeader: string,
    authRegCreds: string,
    customCA: string,
    syncJobPodTemplate: any,
    registrySecrets: string[],
    ociRepositories: string[],
    skipTLS: boolean,
    passCredentials: boolean,
    filter?: IAppRepositoryFilter,
  ) {
    const { data } = await axiosWithAuth.post<ICreateAppRepositoryResponse>(
      url.backend.apprepositories.create(cluster, namespace),
      {
        appRepository: {
          name,
          repoURL,
          authHeader,
          authRegCreds,
          type,
          description,
          customCA,
          syncJobPodTemplate,
          registrySecrets,
          ociRepositories,
          tlsInsecureSkipVerify: skipTLS,
          passCredentials: passCredentials,
          filterRule: filter,
        },
      },
    );
    return data;
  }

  public static async validate(
    cluster: string,
    namespace: string,
    repoURL: string,
    type: string,
    authHeader: string,
    authRegCreds: string,
    customCA: string,
    ociRepositories: string[],
    skipTLS: boolean,
    passCredentials: boolean,
  ) {
    const { data } = await axiosWithAuth.post<any>(
      url.backend.apprepositories.validate(cluster, namespace),
      {
        appRepository: {
          repoURL,
          type,
          authHeader,
          authRegCreds,
          customCA,
          ociRepositories,
          tlsInsecureSkipVerify: skipTLS,
          passCredentials: passCredentials,
        },
      },
    );
    return data;
  }
}
