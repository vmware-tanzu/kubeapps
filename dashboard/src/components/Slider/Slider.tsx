// Copyright 2019-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

// CODE ADAPTED FROM: https://sghall.github.io/react-compound-slider/#/slider-demos/horizontal
import React from "react";
import { Handles, Rail, Slider as ReactSlider, Tracks } from "react-compound-slider";
import { Handle, Track } from "./components"; // example render components

const sliderStyle: React.CSSProperties = {
  margin: "1.2em",
  position: "relative",
  width: "90%",
};

const railStyle: React.CSSProperties = {
  position: "absolute",
  width: "100%",
  height: 14,
  borderRadius: 7,
  cursor: "pointer",
  backgroundColor: "rgb(155,155,155)",
};

export interface ISliderProps {
  min: number;
  max: number;
  default: number;
  step: number;
  values: number;
  sliderStyle?: React.CSSProperties;
  onChange: (values: readonly number[]) => void;
  onUpdate: (values: readonly number[]) => void;
}

class Slider extends React.Component<ISliderProps> {
  public render() {
    const { min, max, step, values, onUpdate, onChange } = this.props;
    const domain = [min, max];
    return (
      <ReactSlider
        mode={1}
        step={step}
        domain={domain}
        rootStyle={{
          ...sliderStyle,
          ...this.props.sliderStyle,
        }}
        onUpdate={onUpdate}
        onChange={onChange}
        values={[values]}
      >
        <Rail>{({ getRailProps }) => <div style={railStyle} {...getRailProps()} />}</Rail>
        <Handles>
          {({ handles, getHandleProps }) => (
            <div className="slider-handles">
              {handles.map(handle => (
                <Handle
                  key={handle.id}
                  handle={handle}
                  domain={domain}
                  getHandleProps={getHandleProps}
                />
              ))}
            </div>
          )}
        </Handles>
        <Tracks right={false}>
          {({ tracks, getTrackProps }) => (
            <div className="slider-tracks">
              {tracks.map(({ id, source, target }) => (
                <Track key={id} source={source} target={target} getTrackProps={getTrackProps} />
              ))}
            </div>
          )}
        </Tracks>
      </ReactSlider>
    );
  }
}

export default Slider;
