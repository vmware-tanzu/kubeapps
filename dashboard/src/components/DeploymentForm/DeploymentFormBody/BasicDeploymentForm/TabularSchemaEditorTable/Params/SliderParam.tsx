// Copyright 2019-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import { CdsControlMessage } from "@cds/react/forms";
import { CdsInput } from "@cds/react/input";
import { CdsRange } from "@cds/react/range";
import Column from "components/js/Column";
import Row from "components/js/Row";
import { isEmpty, toNumber } from "lodash";
import { useState } from "react";
import { validateValuesSchema } from "shared/schema";
import { IAjvValidateResult, IBasicFormParam } from "shared/types";
import { basicFormsDebounceTime, getOptionalMin, getStringValue } from "shared/utils";

export interface ISliderParamProps {
  id: string;
  label: string;
  param: IBasicFormParam;
  unit: string;
  handleBasicFormParamChange: (
    p: IBasicFormParam,
  ) => (e: React.FormEvent<HTMLInputElement>) => void;
}

export default function SliderParam(props: ISliderParamProps) {
  const { handleBasicFormParamChange, id, label, param } = props;

  const initCurrentValue = () =>
    toNumber(param.currentValue) ||
    toNumber(param.exclusiveMinimum) ||
    toNumber(param.minimum) ||
    0;

  const [validated, setValidated] = useState<IAjvValidateResult>();
  const [currentValue, setCurrentValue] = useState(initCurrentValue());

  const [isValueModified, setIsValueModified] = useState(false);
  const [timeout, setThisTimeout] = useState({} as NodeJS.Timeout);

  const step =
    param?.multipleOf || (param.schema?.type === "number" || param?.type === "number" ? 0.5 : 1);

  const onChange = (e: React.FormEvent<HTMLInputElement>) => {
    setCurrentValue(toNumber(e.currentTarget.value));
    setIsValueModified(toNumber(e.currentTarget.value) !== toNumber(param.currentValue));

    // twofold validation: using the json schema (with ajv) and the html5 validation
    setValidated(validateValuesSchema(e.currentTarget.value, param.schema));
    e.currentTarget?.reportValidity();

    // Gather changes before submitting
    clearTimeout(timeout);
    const func = handleBasicFormParamChange(param);
    // The reference to target get lost, so we need to keep a copy
    const targetCopy = {
      currentTarget: {
        value: getStringValue(e.currentTarget?.value),
        type: e.currentTarget?.type,
      },
    } as React.FormEvent<HTMLInputElement>;
    setThisTimeout(setTimeout(() => func(targetCopy), basicFormsDebounceTime));
  };

  const unsavedMessage = isValueModified ? "Unsaved" : "";
  const isModified =
    isValueModified ||
    (param.currentValue !== param.defaultValue && param.currentValue !== param.deployedValue);

  const renderControlMsg = () =>
    !validated?.valid && !isEmpty(validated?.errors) ? (
      <>
        <CdsControlMessage status="error">
          {unsavedMessage}
          <br />
          {validated?.errors?.map(e => e.message).join(", ")}
        </CdsControlMessage>
        <br />
      </>
    ) : (
      <CdsControlMessage>{unsavedMessage}</CdsControlMessage>
    );

  const input = (
    <CdsRange>
      <input
        required={param.isRequired}
        disabled={param.readOnly}
        min={getOptionalMin(param.exclusiveMinimum, param.minimum)}
        max={getOptionalMin(param.exclusiveMaximum, param.maximum)}
        aria-label={label}
        id={id + "_range"}
        type="range"
        step={step}
        onChange={onChange}
        value={currentValue}
      />
      {renderControlMsg()}
    </CdsRange>
  );

  const inputText = (
    <div className="self-center">
      <CdsInput className={isModified ? "bolder" : ""}>
        <input
          required={param.isRequired}
          disabled={param.readOnly}
          min={getOptionalMin(param.exclusiveMinimum, param.minimum)}
          max={getOptionalMin(param.exclusiveMaximum, param.maximum)}
          aria-label={label}
          id={id + "_text"}
          type="number"
          step={step}
          onChange={onChange}
          value={currentValue}
        />
      </CdsInput>
    </div>
  );

  return (
    <Row>
      <Column span={4}>{inputText}</Column>
      <Column span={6}>{input}</Column>
    </Row>
  );
}
