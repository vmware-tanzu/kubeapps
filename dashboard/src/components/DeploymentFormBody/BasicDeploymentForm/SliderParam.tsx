// Copyright 2019-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import { useEffect, useState } from "react";
import { IBasicFormParam } from "shared/types";
import Slider from "../../Slider";

export interface ISliderParamProps {
  id: string;
  label: string;
  param: IBasicFormParam;
  unit: string;
  min: number;
  max: number;
  step: number;
  handleBasicFormParamChange: (
    p: IBasicFormParam,
  ) => (e: React.FormEvent<HTMLInputElement>) => void;
}

export interface ISliderParamState {
  value: number;
}

function toNumber(value: string | number) {
  // Force to return a Number from a string removing any character that is not a digit
  return typeof value === "number" ? value : Number(value.replace(/[^\d.]/g, ""));
}

function getDefaultValue(min: number, value?: string) {
  return (value && toNumber(value)) || min;
}

function SliderParam({
  id,
  label,
  param,
  unit,
  min,
  max,
  step,
  handleBasicFormParamChange,
}: ISliderParamProps) {
  const [value, setValue] = useState(getDefaultValue(min, param.value));

  useEffect(() => {
    setValue(getDefaultValue(min, param.value));
  }, [param, min]);

  const handleParamChange = (newValue: number) => {
    handleBasicFormParamChange(param)({
      currentTarget: {
        value: param.type === "string" ? `${newValue}${unit}` : newValue,
      },
    } as React.FormEvent<HTMLInputElement>);
  };

  // onChangeSlider is run when the slider is dropped at one point
  // at that point we update the parameter
  const onChangeSlider = (values: readonly number[]) => {
    handleParamChange(values[0]);
  };

  // onUpdateSlider is run when dragging the slider
  // we just update the state here for a faster response
  const onUpdateSlider = (values: readonly number[]) => {
    setValue(values[0]);
  };

  const onChangeInput = (e: React.FormEvent<HTMLInputElement>) => {
    const numberValue = toNumber(e.currentTarget.value);
    setValue(numberValue);
    handleParamChange(numberValue);
  };

  return (
    <div>
      <label htmlFor={id}>
        <span className="centered deployment-form-label deployment-form-label-text-param">
          {label}
        </span>
        <div className="slider-block">
          <div className="slider-content">
            <Slider
              // If the parameter defines a minimum or maximum, maintain those
              min={Math.min(param.minimum || min, value)}
              max={Math.max(param.maximum || max, value)}
              step={step || 1}
              default={value}
              onChange={onChangeSlider}
              onUpdate={onUpdateSlider}
              values={value}
              sliderStyle={{ width: "100%", margin: "1.2em 0 1.2em 0" }}
            />
          </div>
          <div className="slider-input-and-unit">
            <input
              className="slider-input clr-input"
              id={id}
              onChange={onChangeInput}
              value={value}
            />
            <span className="margin-l-normal">{unit}</span>
          </div>
        </div>
        {param.description && <span className="description">{param.description}</span>}
      </label>
    </div>
  );
}

export default SliderParam;
