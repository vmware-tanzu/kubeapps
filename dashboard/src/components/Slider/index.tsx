// Copyright 2019-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import Slider from "./Slider";

export default Slider;
