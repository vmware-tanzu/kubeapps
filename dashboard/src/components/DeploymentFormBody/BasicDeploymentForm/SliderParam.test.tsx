// Copyright 2019-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import { shallow } from "enzyme";
import React from "react";
import { IBasicFormParam } from "shared/types";
import Slider from "../../Slider";
import SliderParam from "./SliderParam";

const defaultProps = {
  id: "disk",
  label: "Disk Size",
  handleBasicFormParamChange: jest.fn(() => jest.fn()),
  min: 1,
  max: 100,
  step: 1,
  unit: "Gi",
};

const params: IBasicFormParam[] = [
  {
    value: "10Gi",
    type: "string",
    path: "disk",
  },
  {
    value: 10,
    type: "integer",
    path: "disk",
  },
  {
    value: 10.0,
    type: "number",
    path: "disk",
  },
];

it("renders a disk size param with a default value", () => {
  params.forEach(param => {
    const wrapper = shallow(<SliderParam {...defaultProps} param={param} />);
    expect(wrapper.find(Slider).prop("values")).toBe(10);
    expect(wrapper).toMatchSnapshot();
  });
});

describe("when changing the slide", () => {
  it("changes the value of the string param", () => {
    params.forEach(param => {
      const cloneParam = { ...param } as IBasicFormParam;
      const expected = param.type === "string" ? "20Gi" : 20;

      const handleBasicFormParamChange = jest.fn(() => {
        cloneParam.value = expected;
        return jest.fn();
      });

      const wrapper = shallow(
        <SliderParam
          {...defaultProps}
          param={cloneParam}
          handleBasicFormParamChange={handleBasicFormParamChange}
        />,
      );

      expect(wrapper.find(Slider).prop("values")).toBe(10);

      const slider = wrapper.find(Slider);
      (slider.prop("onChange") as (values: number[]) => void)([20]);

      expect(cloneParam.value).toBe(expected);
      expect(handleBasicFormParamChange.mock.calls[0]).toEqual([
        { value: expected, type: param.type, path: param.path },
      ]);
    });
  });

  it("changes the value of the string param without unit", () => {
    params.forEach(param => {
      const cloneParam = { ...param } as IBasicFormParam;
      const expected = param.type === "string" ? "20" : 20;

      const handleBasicFormParamChange = jest.fn(() => {
        cloneParam.value = expected;
        return jest.fn();
      });

      const wrapper = shallow(
        <SliderParam
          {...defaultProps}
          param={cloneParam}
          unit=""
          handleBasicFormParamChange={handleBasicFormParamChange}
        />,
      );

      expect(wrapper.find(Slider).prop("values")).toBe(10);

      const slider = wrapper.find(Slider);
      (slider.prop("onChange") as (values: number[]) => void)([20]);

      expect(cloneParam.value).toBe(expected);
      expect(handleBasicFormParamChange.mock.calls[0]).toEqual([
        { value: expected, type: param.type, path: param.path },
      ]);
    });
  });

  it("changes the value of the string param with the step defined", () => {
    params.forEach(param => {
      const cloneProps = { ...defaultProps, step: 10 };
      const cloneParam = { ...param } as IBasicFormParam;
      const expected = param.type === "string" ? "20Gi" : 20;

      const handleBasicFormParamChange = jest.fn(() => {
        cloneParam.value = expected;
        return jest.fn();
      });

      const wrapper = shallow(
        <SliderParam
          {...cloneProps}
          param={cloneParam}
          handleBasicFormParamChange={handleBasicFormParamChange}
        />,
      );

      expect(wrapper.find(Slider).prop("values")).toBe(10);

      const slider = wrapper.find(Slider);
      (slider.prop("onChange") as (values: number[]) => void)([2]);

      expect(cloneParam.value).toBe(expected);
      expect(handleBasicFormParamChange.mock.calls[0]).toEqual([
        { value: expected, type: param.type, path: param.path },
      ]);
    });
  });
});

it("updates state but does not change param value during slider update (only when dropped in a point)", () => {
  params.forEach(param => {
    const handleBasicFormParamChange = jest.fn();
    const wrapper = shallow(
      <SliderParam
        {...defaultProps}
        param={param}
        handleBasicFormParamChange={handleBasicFormParamChange}
      />,
    );
    expect(wrapper.find(Slider).prop("values")).toBe(10);

    const slider = wrapper.find(Slider);
    (slider.prop("onUpdate") as (values: number[]) => void)([20]);

    expect(wrapper.find(Slider).prop("values")).toBe(20);
    expect(handleBasicFormParamChange).not.toHaveBeenCalled();
  });
});

describe("when changing the value in the input", () => {
  it("parses a number and forwards it", () => {
    params.forEach(param => {
      const valueChange = jest.fn();
      const handleBasicFormParamChange = jest.fn(() => valueChange);
      const wrapper = shallow(
        <SliderParam
          {...defaultProps}
          param={param}
          handleBasicFormParamChange={handleBasicFormParamChange}
        />,
      );
      expect(wrapper.find(Slider).prop("values")).toBe(10);

      const input = wrapper.find("input#disk");
      const event = { currentTarget: { value: "20" } } as React.FormEvent<HTMLInputElement>;
      (input.prop("onChange") as (e: React.FormEvent<HTMLInputElement>) => void)(event);

      expect(wrapper.find(Slider).prop("values")).toBe(20);

      const expected = param.type === "string" ? "20Gi" : 20;
      expect(valueChange.mock.calls[0]).toEqual([{ currentTarget: { value: expected } }]);
    });
  });

  it("parses a number and forwards it without unit", () => {
    params.forEach(param => {
      const valueChange = jest.fn();
      const handleBasicFormParamChange = jest.fn(() => valueChange);
      const wrapper = shallow(
        <SliderParam
          {...defaultProps}
          unit=""
          param={param}
          handleBasicFormParamChange={handleBasicFormParamChange}
        />,
      );
      expect(wrapper.find(Slider).prop("values")).toBe(10);

      const input = wrapper.find("input#disk");
      const event = { currentTarget: { value: "20" } } as React.FormEvent<HTMLInputElement>;
      (input.prop("onChange") as (e: React.FormEvent<HTMLInputElement>) => void)(event);

      expect(wrapper.find(Slider).prop("values")).toBe(20);

      const expected = param.type === "string" ? "20" : 20;
      expect(valueChange.mock.calls[0]).toEqual([{ currentTarget: { value: expected } }]);
    });
  });

  it("ignores values in the input that are not digits", () => {
    params.forEach(param => {
      const valueChange = jest.fn();
      const handleBasicFormParamChange = jest.fn(() => valueChange);
      const wrapper = shallow(
        <SliderParam
          {...defaultProps}
          param={param}
          handleBasicFormParamChange={handleBasicFormParamChange}
        />,
      );
      expect(wrapper.find(Slider).prop("values")).toBe(10);

      const input = wrapper.find("input#disk");
      const event = { currentTarget: { value: "foo20*#@$" } } as React.FormEvent<HTMLInputElement>;
      (input.prop("onChange") as (e: React.FormEvent<HTMLInputElement>) => void)(event);

      expect(wrapper.find(Slider).prop("values")).toBe(20);

      const expected = param.type === "string" ? "20Gi" : 20;
      expect(valueChange.mock.calls[0]).toEqual([{ currentTarget: { value: expected } }]);
    });
  });

  it("accept decimal values", () => {
    params.forEach(param => {
      const valueChange = jest.fn();
      const handleBasicFormParamChange = jest.fn(() => valueChange);
      const wrapper = shallow(
        <SliderParam
          {...defaultProps}
          param={param}
          handleBasicFormParamChange={handleBasicFormParamChange}
        />,
      );
      expect(wrapper.find(Slider).prop("values")).toBe(10);

      const input = wrapper.find("input#disk");
      const event = { currentTarget: { value: "20.5" } } as React.FormEvent<HTMLInputElement>;
      (input.prop("onChange") as (e: React.FormEvent<HTMLInputElement>) => void)(event);

      expect(wrapper.find(Slider).prop("values")).toBe(20.5);

      const expected = param.type === "string" ? "20.5Gi" : 20.5;
      expect(valueChange.mock.calls[0]).toEqual([{ currentTarget: { value: expected } }]);
    });
  });

  it("modifies the max value of the slider if the input is greater than 100", () => {
    params.forEach(param => {
      const valueChange = jest.fn();
      const handleBasicFormParamChange = jest.fn(() => valueChange);
      const wrapper = shallow(
        <SliderParam
          {...defaultProps}
          param={param}
          handleBasicFormParamChange={handleBasicFormParamChange}
        />,
      );
      expect(wrapper.find(Slider).prop("values")).toBe(10);

      const input = wrapper.find("input#disk");
      const event = { currentTarget: { value: "200" } } as React.FormEvent<HTMLInputElement>;
      (input.prop("onChange") as (e: React.FormEvent<HTMLInputElement>) => void)(event);

      expect(wrapper.find(Slider).prop("values")).toBe(200);
      const slider = wrapper.find(Slider);
      expect(slider.prop("max")).toBe(200);
    });
  });
});

it("uses the param minimum and maximum if defined", () => {
  params.forEach(param => {
    const clonedParam = { ...param } as IBasicFormParam;
    clonedParam.minimum = 5;
    clonedParam.maximum = 50;

    const wrapper = shallow(<SliderParam {...defaultProps} param={clonedParam} />);

    const slider = wrapper.find(Slider);
    expect(slider.prop("min")).toBe(5);
    expect(slider.prop("max")).toBe(50);
  });
});

it("defaults to the min if the value is undefined", () => {
  params.forEach(param => {
    const cloneParam = { ...param } as IBasicFormParam;
    cloneParam.value = undefined;

    const wrapper = shallow(<SliderParam {...defaultProps} param={cloneParam} min={5} />);
    expect(wrapper.find(Slider).prop("values")).toBe(5);
  });
});
