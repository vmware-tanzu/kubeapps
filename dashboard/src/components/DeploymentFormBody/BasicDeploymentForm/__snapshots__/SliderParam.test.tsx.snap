// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a disk size param with a default value 1`] = `
<div>
  <label
    htmlFor="disk"
  >
    <span
      className="centered deployment-form-label deployment-form-label-text-param"
    >
      Disk Size
    </span>
    <div
      className="slider-block"
    >
      <div
        className="slider-content"
      >
        <Slider
          default={10}
          max={100}
          min={1}
          onChange={[Function]}
          onUpdate={[Function]}
          sliderStyle={
            Object {
              "margin": "1.2em 0 1.2em 0",
              "width": "100%",
            }
          }
          step={1}
          values={10}
        />
      </div>
      <div
        className="slider-input-and-unit"
      >
        <input
          className="slider-input clr-input"
          id="disk"
          onChange={[Function]}
          value={10}
        />
        <span
          className="margin-l-normal"
        >
          Gi
        </span>
      </div>
    </div>
  </label>
</div>
`;

exports[`renders a disk size param with a default value 2`] = `
<div>
  <label
    htmlFor="disk"
  >
    <span
      className="centered deployment-form-label deployment-form-label-text-param"
    >
      Disk Size
    </span>
    <div
      className="slider-block"
    >
      <div
        className="slider-content"
      >
        <Slider
          default={10}
          max={100}
          min={1}
          onChange={[Function]}
          onUpdate={[Function]}
          sliderStyle={
            Object {
              "margin": "1.2em 0 1.2em 0",
              "width": "100%",
            }
          }
          step={1}
          values={10}
        />
      </div>
      <div
        className="slider-input-and-unit"
      >
        <input
          className="slider-input clr-input"
          id="disk"
          onChange={[Function]}
          value={10}
        />
        <span
          className="margin-l-normal"
        >
          Gi
        </span>
      </div>
    </div>
  </label>
</div>
`;

exports[`renders a disk size param with a default value 3`] = `
<div>
  <label
    htmlFor="disk"
  >
    <span
      className="centered deployment-form-label deployment-form-label-text-param"
    >
      Disk Size
    </span>
    <div
      className="slider-block"
    >
      <div
        className="slider-content"
      >
        <Slider
          default={10}
          max={100}
          min={1}
          onChange={[Function]}
          onUpdate={[Function]}
          sliderStyle={
            Object {
              "margin": "1.2em 0 1.2em 0",
              "width": "100%",
            }
          }
          step={1}
          values={10}
        />
      </div>
      <div
        className="slider-input-and-unit"
      >
        <input
          className="slider-input clr-input"
          id="disk"
          onChange={[Function]}
          value={10}
        />
        <span
          className="margin-l-normal"
        >
          Gi
        </span>
      </div>
    </div>
  </label>
</div>
`;
